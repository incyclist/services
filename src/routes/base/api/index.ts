import {EventLogger} from 'gd-eventlog'
import { RouteApiDescription, RouteApiDetail, RouteDescriptionQuery } from "./types";
import { AxiosInstance } from "axios";
import { DEFAULT_ROUTE_API, NO_CACHE, ROUTE_API } from './consts'
import { IncyclistRestApiClient } from '../../../api';
import { useUserSettings } from '../../../settings';
import { RoutePoint } from '../types';

export default class IncyclistRoutesApi { 

    protected static _instance

    static getInstance():IncyclistRoutesApi{
        if (!IncyclistRoutesApi._instance) 
        IncyclistRoutesApi._instance = new IncyclistRoutesApi()
        
        return IncyclistRoutesApi._instance
    }

    protected logger: EventLogger
    protected api: AxiosInstance

    constructor() {
        this.logger = new EventLogger('RouteListApi')
    }

    protected logError( err:Error, fn:string, logInfo?) {
        const args = logInfo || {}
        this.logger.logEvent( {message:'Error', error:err.message, fn, ...args})
    }

    protected getApi():AxiosInstance {
        if (!this.api) {
            this.api = IncyclistRestApiClient.getClient()
            return this.api
        }
        return this.api
    }


    protected getBaseUrl() {
        
        return useUserSettings().get(ROUTE_API,DEFAULT_ROUTE_API)
    }

    async getRouteDescriptions(query:RouteDescriptionQuery, throws:boolean = false): Promise<Array<RouteApiDescription>> {
        const {type='gpx',category} = query;

        let url  = (type === 'gpx' ) ? `?private=false` : `/?type=${type}`;
        if (category)
            url += "&category=${category}"

        try {
            const res = await this._get( url, NO_CACHE )       
            return res.data;           
        }
        catch(err) {
            if (throws)
                throw(err)

            this.logError(err,'getRouteDescriptions', {query})
            return undefined
        }
    }

    static verify(route) {
        if (route['decoded'] ) {
            route.points =route['decoded']
            delete route['decoded']

            IncyclistRoutesApi.fixMissingRouteDistances(route.points)
        }

    }

    protected static fixMissingRouteDistances(points:Array<RoutePoint>) {
        let routeDistance = 0
        points.forEach( (p)=> {
            if (!p.routeDistance)
                p.routeDistance = routeDistance
            routeDistance = p.routeDistance+p['distance']
        })

    }

    async getRouteDetails( routeId: string): Promise<RouteApiDetail> {

        try {
            const res = await this._get( `/${routeId}` )  
            IncyclistRoutesApi.verify(res.data)
            return res.data;           
        }
        catch(err) {
            this.logError(err,'getRouteDetails', {routeId})
            return undefined
        }
    }

    async getRoutePreview( routeId: string): Promise<string> {

        try {
            const res = await this._get( `/${routeId}/preview` )              
            return res.data?.url;           
        }
        catch(err) {
            return undefined
        }
    }

    async reload():Promise<void> {
        try {
            await this._get( `/reload` )                   
        }
        catch(err) {
            this.logError(err,'reload')
        }

    }

    protected async _get(url:string, ...args) {
        const api = this.getApi()
        const baseUrl = this.getBaseUrl()
        
        return await api.get( baseUrl+url, ...args )       
    }


}




