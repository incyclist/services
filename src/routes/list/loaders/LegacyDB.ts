import { JSONObject } from "../../../api"
import { Singleton } from "../../../base/types"
import { Observer, PromiseObserver } from "../../../base/types/observer"
import { valid } from "../../../utils/valid"
import { RouteInfo } from "../../base/types"
import { RouteDBApiDescription } from "./types"
import { DBLoader } from "./DBLoader"

@Singleton
export class RoutesLegacyDbLoader extends DBLoader<RouteDBApiDescription>{
    
    protected loadObserver: Observer
    protected saveObserver: PromiseObserver<void>
    

    constructor () {
        super()
        this.routeDescriptions = []        
    }


    async save(): Promise<void> {
        // not required - routes should only be saved from DBLoader
        return;
    } 

    protected getIsLocal(descr:RouteDBApiDescription):boolean {
        const {type,category} = descr
        if (type==='video') {
            return !valid(category) || category==='Imported'
        }
        if (type==='gpx') {
            return category==='personal'
        }        
    }


    protected buildRouteInfo(descr:RouteDBApiDescription):RouteInfo {
        const { id,legacyId,title,localizedTitle,country,distance,elevation, category,provider, video, points,type,routeHash, originalName} = descr
        
        const data:RouteInfo = { id,legacyId,title,localizedTitle,country,distance,elevation,provider,category,routeHash,originalName}

        data.hasVideo = false;
        data.hasGpx = type==='gpx'
        data.isLocal = this.getIsLocal(descr)
        

        this.updateRouteCountry(data,{descr})
        this.updateRouteTitle(data,{descr})
        // Todo: previewImg (could be generated from video/streetview)

        if (points) data.hasGpx = true;

        if (category?.toLowerCase()==='demo') 
            data.isDemo = true;

        if (video) {
            data.hasVideo = true;
            if (video.format) data.videoFormat = video.format
            if (video.url && !data.videoUrl) data.videoUrl = video.url            
        }

        return data;
    }

    protected buildRouteDBInfo():RouteDBApiDescription {
        throw new Error('not implemented')
    }

    protected isCompleted(): boolean {
        // ensure that loadDetails is not called and 'route.added' events are emitted
        return true        
    }

    protected checkLegacy(descr:RouteDBApiDescription ):RouteDBApiDescription {
        if (descr.routeId) {
            descr.legacyId = descr.id
            descr.id = descr.routeId
            delete descr.routeId
        }
        return descr
    }

    protected async loadDescriptions():Promise<Array<RouteDBApiDescription>> {

        const add = (data:JSONObject, type:'video'|'gpx') => {
            try {
                const routes = data as unknown as Array<RouteDBApiDescription>                
                return routes.map( r=> {
                    r.type = type;
                    return r
                })
            }
            catch(err) {
                return null;
            }

        }

        const videos = this.getVideosRepo().read('routes').then( data=>add(data,'video'))
        const routes = this.getRoutesRepo().read('routes').then( data=>add(data,'gpx'))

        const descriptions:Array<RouteDBApiDescription> = []

        const res = await Promise.allSettled( [videos,routes]) 
        res.forEach( p => {
            if (p.status==='fulfilled' && p.value) {                   
                const data = p.value.map( d=>this.checkLegacy(d))
                
                descriptions.push( ...data)
            }
        })
       
        return descriptions;
    }

    protected async loadDetails():Promise<void> {
        // Legacy DB is only used to load descriptions
        // details should be loaded via DB Loader
        throw new Error('not implemented')
    }

}
