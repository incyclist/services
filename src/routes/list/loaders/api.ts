import { Observer, PromiseObserver } from "../../../base/types/observer";
import IncyclistRoutesApi from "../../base/api";
import { RouteApiDescription, RouteApiDetail } from "../../base/api/types";
import { Route } from "../../base/model/route";
import { RouteInfo } from "../../base/types";
import { RoutesDbLoader } from "./db";
import { Loader } from "./types";
import { LoadDetailsTargets } from "./types";

const valid = (v) => (v!==undefined && v!==null)

export class RoutesApiLoader extends Loader<RouteApiDescription> {

    protected api: IncyclistRoutesApi
    protected loadObserver: Observer
    protected saveObserver: PromiseObserver<void>

    constructor () {
        super()
        this.api = IncyclistRoutesApi.getInstance()        
    }


    load():Observer {
        // no concurrent loads
        if (this.loadObserver)
            return this.loadObserver

        const enrichApi = (v:RouteApiDescription,type)=> ({type, ...v })
        const api = [
            this.api.getRouteDescriptions({type:'gpx'}).then( v=> v.map(e=>enrichApi(e,'gpx'))) ,
            this.api.getRouteDescriptions({type:'video'}).then( v=> v.map(e=>enrichApi(e,'video')))
        ]

        this.loadObserver = new Observer()

        Promise.allSettled( api).then (async res => {

            // join lists
            const items: Array<{route:Route, added:boolean}> = []

            // for loop as we have asyncs here
            for ( let i=0; i<res.length;i++) {
                const p = res[i]

                if (p.status==='fulfilled') {  
                    for (let j=0;j<p.value.length;j++) {
                        const descr = p.value[j]

                    
                        let description;

                        const existing = this.getDescriptionFromDB(descr.routeId||descr.id)
                        const {isDeleted=false} = descr

                        if (existing 
                              && ((existing.version||0)>=(descr.version||0)) 
                              && (!isDeleted || (isDeleted && existing.isDownloaded))
                            ) {

                            const details = await this.getDetailsFromDB(existing.id)
                            if (details) {
                                continue;
                            }
                            description = existing
                            
                        }
                        else {
                            description = this.buildRouteInfo(descr)
                        }

                        if (!description) {
                            continue
                        }


                        
                        const route = new Route( description)

                        const isComplete = this.isCompleted(route)
                        
                        if (isComplete) {
                            this.verifyRouteHash(route)
                            this.emitRouteAdded(route)
                        }

                        if (!existing) {
                            items.push({route,added:isComplete})
                        }
                        else {
                            this.save(route,true)
                        }

                    }
                }
            }

            await this.loadDetails(items)
            
            //delete this.loadObserver
        });

        return this.loadObserver
        
    }

    stopLoad() {
        delete this.loadObserver
    }

    protected getDescriptionFromDB(id:string) {
        const db = new RoutesDbLoader()
        return db.getDescription(id)

    }
    protected async getDetailsFromDB(id:string):Promise<RouteApiDetail> {

        const db = new RoutesDbLoader()        
        return await db.getDetails(id)

    }

    async save(route:Route,enforcedWriteDetails:boolean=false):Promise<void> {
        const db = new RoutesDbLoader()
        await db.save(route,enforcedWriteDetails)
    }


    protected isCompleted(route:Route):boolean {
        const descr = route.description

        if (descr.hasVideo) {
            return valid(descr.points) && ( valid(descr.videoUrl)) || (descr.requiresDownload && valid(descr.downloadUrl))
        }
        else {
            return valid(descr.points)
        }
    }

    protected isPreviewMissing(route:Route) {
        const descr = route.description
        return descr?.hasVideo && !descr?.previewUrl
    }

    protected async loadDetailsFromRepo( items:LoadDetailsTargets):Promise<LoadDetailsTargets>{

        // all missing should be loaded from server
        const promises = items.map ( i => this.getDetailsFromDB (i.route.description.id).then( details => { this.addDetails(i.route,details)}) )

        const res = await Promise.allSettled(promises) 

        const {failed} = this.processLoadDetailsResult(res, items, false);
        return failed
        
    }


    async loadDetails( items:LoadDetailsTargets) {


        // first try to load details from repo
        const failed = await this.loadDetailsFromRepo(items)


        // all missing should be loaded from server
        const promises = failed.map ( i => this.getDetails(i.route) )

        const res = await Promise.allSettled(promises) 

        this.processLoadDetailsResult(res, items, true);

    }

    private processLoadDetailsResult(res: PromiseSettledResult<void>[],items: LoadDetailsTargets, save?:boolean):{success:LoadDetailsTargets,  failed: LoadDetailsTargets} {
        const success:LoadDetailsTargets = []
        const failed: LoadDetailsTargets = []

        res.forEach((pr, idx) => {
            if (pr.status === 'fulfilled') success.push(items[idx]);
            else failed.push(items[idx]);
        });

        success.forEach(item => {
            if (item.added)
                this.emitRouteUpdate(item.route);

            else
                this.emitRouteAdded(item.route);


            this.verifyCountry(item.route);
            this.verifyRouteHash(item.route)

            if (save)
                this.save(item.route, true);

        });

        return {success,failed}
    }

    protected async getDetails(route:Route):Promise<void> {

        const details = await this.api.getRouteDetails(route.description.legacyId||route.description.id)
        this.addDetails(route,details)

            
    }

    protected buildRouteInfo( descr:RouteApiDescription, isLocal?:boolean): RouteInfo {
        const { id,routeId,title,localizedTitle,country,distance,elevation, category,provider, video, points,previewUrl,routeHash,version,isDeleted} = descr
        
        const data:RouteInfo = { title,localizedTitle,country,distance,elevation,provider,category,previewUrl,routeHash,version,isDeleted}

        data.hasVideo = false;
        data.hasGpx = false;
        data.isLocal = isLocal||false
        data.id = routeId || id
        data.legacyId = routeId? id: undefined


        this.updateRouteCountry(data,{descr})
        this.updateRouteTitle(data,{descr})
        // Todo: previewImg (could be generated from video/streetview)

        if (points) data.hasGpx = true;

        if (category?.toLowerCase()==='demo') 
            data.isDemo = true;

        if (video) {
            data.hasVideo = true;
            if (video.format) data.videoFormat = video.format
            if (video.url && !data.videoUrl) data.videoUrl = video.url            
        }

        return data;
    }

 
}