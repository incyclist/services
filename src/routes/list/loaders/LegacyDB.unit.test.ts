import routesData from '../../../../__tests__/data/db/routes.json';
import videosData from '../../../../__tests__/data/db/videos.json';

import { JsonRepository } from '../../../api';
import { Route } from "../../base/model/route";
import { RoutesLegacyDbLoader } from "./LegacyDB";



describe('LegacyDBLoader',()=>{
    describe('load',()=>{

  
        let loader;
        beforeEach( ()=>{
            loader = new RoutesLegacyDbLoader()                    
            
        })

        afterEach( ()=>{
            RoutesLegacyDbLoader['_instance'] = undefined
        })

        test('parsing descriptions',async ()=>{
            const routes:Array<Route> = []


            const videosRepo = JsonRepository.create('videos')
            const routesRepo = JsonRepository.create('routes')
    
            videosRepo.read = jest.fn().mockResolvedValue(videosData)
            routesRepo.read = jest.fn().mockResolvedValue(routesData)

            const routeIdCnt = videosData.filter( r=>r.routeId!==undefined)?.length

            await new Promise(done=>{
                loader.load()
                    .on('route.added', route=>{routes.push(route)})
                    .on('done',done)}
                )

            const legacyIdCnt = routes.filter( r=>r.description.legacyId!==undefined)?.length
            expect(legacyIdCnt).toBe(routeIdCnt)
            expect(routes.length).toBe(34)

            expect(routes.map(r=>r.description)).toMatchSnapshot()
        })

    })

})