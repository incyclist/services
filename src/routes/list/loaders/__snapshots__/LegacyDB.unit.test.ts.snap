// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LegacyDBLoader load parsing descriptions 1`] = `
[
  {
    "category": "Imported",
    "country": "de",
    "distance": 11722,
    "elevation": 201.9054766945829,
    "hasGpx": false,
    "hasVideo": true,
    "id": "20ef79b5-6ce0-4140-9d3e-3aa9ad728f36",
    "isLocal": true,
    "legacyId": undefined,
    "localizedTitle": undefined,
    "provider": {
      "logo": "https://videos.incyclist.com/reallifevideo.svg",
      "name": "RealLifeVideo",
      "url": "https://www.reallifevideo.de/index.php",
    },
    "routeHash": "4287767686b6327b8284fab983c6621b",
    "title": "Arnbach",
    "videoFormat": "mp4",
  },
  {
    "category": "Imported",
    "country": "FR",
    "distance": 11992.295979674931,
    "elevation": 154.35999999999996,
    "hasGpx": false,
    "hasVideo": true,
    "id": "e8b20d962bc5404db7ddb76ad877132c",
    "isLocal": true,
    "legacyId": "e28066cd-5c00-462f-97a5-07ce2e8e1aae",
    "localizedTitle": undefined,
    "provider": undefined,
    "routeHash": "0a7566592f510c8c0482a3fa8e2c975a",
    "title": "Menglonnais",
    "videoFormat": "mp4",
  },
  {
    "category": "Free",
    "country": "IT",
    "distance": 13000,
    "elevation": 771.5589632626577,
    "hasGpx": false,
    "hasVideo": true,
    "id": "8ec98050-9bee-4e76-9a9f-e0c2ab1756f3",
    "isLocal": false,
    "legacyId": undefined,
    "localizedTitle": undefined,
    "provider": undefined,
    "routeHash": "a1951c0af7b35c1812f4fba689b0d4dc",
    "title": "Passo Pordoi",
    "videoFormat": "mp4",
  },
  {
    "category": "Free",
    "country": "GR",
    "distance": 40166,
    "elevation": 722.9826956451793,
    "hasGpx": false,
    "hasVideo": true,
    "id": "b95365e59ac8d783a7847f9144b288dc",
    "isLocal": false,
    "legacyId": "effd6fd1-2661-42ad-ae12-75145ca59be2",
    "localizedTitle": undefined,
    "provider": undefined,
    "routeHash": "6998fa0e143687a6b0687fc9e8fd3c69",
    "title": "Polygyros",
    "videoFormat": "mp4",
    "videoUrl": "https://videos.incyclist.com/GR_Polygyros.mp4",
  },
  {
    "category": "Demo",
    "country": "PT",
    "distance": 3254,
    "elevation": 46.39705132629575,
    "hasGpx": false,
    "hasVideo": true,
    "id": "c20c846e-1415-48c1-ac68-525ce6eefe25",
    "isDemo": true,
    "isLocal": false,
    "legacyId": undefined,
    "localizedTitle": {
      "de": "Alferce",
      "en": "Alferce",
    },
    "provider": {
      "logo": "https://videos.incyclist.com/reallifevideo.svg",
      "name": "RealLifeVideo",
      "url": "https://www.reallifevideo.de/index.php",
    },
    "routeHash": "f0dcb7925f78b24bc30818289e9f14b6",
    "title": "Alferce_Demo",
    "videoFormat": "mp4",
  },
  {
    "category": "Demo",
    "country": "DE",
    "distance": 3229,
    "elevation": 27.74599999999907,
    "hasGpx": false,
    "hasVideo": true,
    "id": "1bcd4de5-ad87-4943-8e76-d85b9bf24360",
    "isDemo": true,
    "isLocal": false,
    "legacyId": undefined,
    "localizedTitle": {
      "de": "Allgäu",
      "en": "Allgäu",
    },
    "provider": undefined,
    "routeHash": "a8d007823c038d41d2f9c990cd32be49",
    "title": "Allgäu_Demo",
    "videoFormat": "mp4",
  },
  {
    "category": "Demo",
    "country": "FR",
    "distance": 4970,
    "elevation": 326.1525847688363,
    "hasGpx": false,
    "hasVideo": true,
    "id": "f15b62d7-7976-42d1-bbe6-eb4884c8b703",
    "isDemo": true,
    "isLocal": false,
    "legacyId": undefined,
    "localizedTitle": {
      "de": "Col du Galibier",
      "en": "Col du Galibier",
    },
    "provider": {
      "logo": "https://videos.incyclist.com/reallifevideo.svg",
      "name": "RealLifeVideo",
      "url": "https://www.reallifevideo.de/index.php",
    },
    "routeHash": "4489fd75d9d9b19c4945f931869e799c",
    "title": "Col du Galibier",
    "videoFormat": "mp4",
  },
  {
    "category": "Demo",
    "country": "IT",
    "distance": 1639,
    "elevation": 112.41200000000003,
    "hasGpx": false,
    "hasVideo": true,
    "id": "9f808dfe-808a-444a-85ce-b166c8f899e0",
    "isDemo": true,
    "isLocal": false,
    "legacyId": undefined,
    "localizedTitle": {
      "de": "Corvara",
      "en": "Corvara",
    },
    "provider": undefined,
    "routeHash": "e88ca3fed9c7ff9f55b2d39403e143ad",
    "title": "CiM-Demo-Corvara",
    "videoFormat": "mp4",
  },
  {
    "category": "Demo",
    "country": "ES",
    "distance": 4449,
    "elevation": 138.01776235662376,
    "hasGpx": false,
    "hasVideo": true,
    "id": "89fa6eda-cee7-4bdb-b1ce-d79e98c84c77",
    "isDemo": true,
    "isLocal": false,
    "legacyId": undefined,
    "localizedTitle": {
      "de": "Mallorca: Alcudia - Sant Elm",
      "en": "Mallorca: Alcudia - Sant Elm",
    },
    "provider": {
      "logo": "https://videos.incyclist.com/reallifevideo.svg",
      "name": "RealLifeVideo",
      "url": "https://www.reallifevideo.de/index.php",
    },
    "routeHash": "0a11b4842229d33c4ef8e83cac44c820",
    "title": "AlcStElm_Demo",
    "videoFormat": "mp4",
  },
  {
    "category": "Demo",
    "country": "ES",
    "distance": 2877,
    "elevation": 15.351999999999995,
    "hasGpx": false,
    "hasVideo": true,
    "id": "0abad56e-c34a-46b1-9ca6-3b80ce74925f",
    "isDemo": true,
    "isLocal": false,
    "legacyId": undefined,
    "localizedTitle": {
      "de": "Menorca West",
      "en": "Menorca West",
    },
    "provider": {
      "logo": "https://videos.incyclist.com/reallifevideo.svg",
      "name": "RealLifeVideo",
      "url": "https://www.reallifevideo.de/index.php",
    },
    "routeHash": "7efd0a4d722635b1af9422a2a32b4ddd",
    "title": "Menorca",
    "videoFormat": "mp4",
  },
  {
    "category": "Demo",
    "country": "IT",
    "distance": 4782,
    "elevation": 2.171999999999997,
    "hasGpx": false,
    "hasVideo": true,
    "id": "e0917134-dd7e-4837-ae8a-6b0e239a2682",
    "isDemo": true,
    "isLocal": false,
    "legacyId": undefined,
    "localizedTitle": {
      "de": "Monto Baldo",
      "en": "Monte Baldo",
    },
    "provider": undefined,
    "routeHash": "c083b9974e9501076a21a1164a5d209f",
    "title": "Monte-Baldo_Demo",
    "videoFormat": "mp4",
  },
  {
    "category": "Demo",
    "country": "DE",
    "distance": 4639,
    "elevation": 20.733891318717355,
    "hasGpx": false,
    "hasVideo": true,
    "id": "3fe5ca2f-5048-4c34-95bc-cb238e8f1e4c",
    "isDemo": true,
    "isLocal": false,
    "legacyId": undefined,
    "localizedTitle": {
      "de": "Queichtalradweg",
      "en": "Queichtal bikeway",
    },
    "provider": {
      "logo": "https://videos.incyclist.com/reallifevideo.svg",
      "name": "RealLifeVideo",
      "url": "https://www.reallifevideo.de/index.php",
    },
    "routeHash": "fb289f79a521c12377b8cef5d2bda57c",
    "title": "Queich_Demo",
    "videoFormat": "mp4",
  },
  {
    "category": "Demo",
    "country": "DE",
    "distance": 3464,
    "elevation": 4.311499107324153,
    "hasGpx": false,
    "hasVideo": true,
    "id": "1da71d03-b339-48ab-8fd8-9a53a4f3975d",
    "isDemo": true,
    "isLocal": false,
    "legacyId": undefined,
    "localizedTitle": {
      "de": "Rheinradweg 2022",
      "en": "Rhine Cycle Path 2022",
    },
    "provider": {
      "logo": "https://videos.incyclist.com/reallifevideo.svg",
      "name": "RealLifeVideo",
      "url": "https://www.reallifevideo.de/index.php",
    },
    "routeHash": "f31b5221a89a9bb53ad61e68e79c26b4",
    "title": "Rheinradweg 2022",
    "videoFormat": "mp4",
  },
  {
    "category": "Demo",
    "country": "DE",
    "distance": 4999,
    "elevation": 9.009253154110212,
    "hasGpx": false,
    "hasVideo": true,
    "id": "cad14755-4db2-4ba6-bb3b-d079f8266241",
    "isDemo": true,
    "isLocal": false,
    "legacyId": undefined,
    "localizedTitle": {
      "de": "Rheinradweg Mittelrhein",
      "en": "Rhine Bikeway",
    },
    "provider": {
      "logo": "https://videos.incyclist.com/reallifevideo.svg",
      "name": "RealLifeVideo",
      "url": "https://www.reallifevideo.de/index.php",
    },
    "routeHash": "e0f7187e408bdfee71c5518f8bf97e9f",
    "title": "MRhein_Demo",
    "videoFormat": "mp4",
  },
  {
    "category": "Demo",
    "country": "NO",
    "distance": 2426,
    "elevation": 207.88899999999984,
    "hasGpx": false,
    "hasVideo": true,
    "id": "3510255c-f2da-41db-b0d0-5272f8a2fdf4",
    "isDemo": true,
    "isLocal": false,
    "legacyId": undefined,
    "localizedTitle": {
      "de": "Trollstigen",
      "en": "Trollstigen",
    },
    "provider": {
      "logo": "https://videos.incyclist.com/reallifevideo.svg",
      "name": "RealLifeVideo",
      "url": "https://www.reallifevideo.de/index.php",
    },
    "routeHash": "0c80ca8a3a5ef0ea65c5c86eb2020bd4",
    "title": "Troll-Demo",
    "videoFormat": "mp4",
  },
  {
    "category": "Free",
    "country": "FR",
    "distance": 22645.541971360508,
    "elevation": 517.6000000000001,
    "hasGpx": false,
    "hasVideo": true,
    "id": "d35cf23e-90d3-4b77-8cb8-b1255b8b6ddf",
    "isLocal": false,
    "legacyId": "a07c62c0-61a0-4f81-b55f-e743b9245006",
    "localizedTitle": undefined,
    "provider": undefined,
    "routeHash": "70fd6b6ae8d29083686c80d4c542b091",
    "title": "Valdrôme 1 - Col de Prémol",
    "videoFormat": "mp4",
  },
  {
    "category": "Free",
    "country": "FR",
    "distance": 17116.539243010033,
    "elevation": 408.60000000000036,
    "hasGpx": false,
    "hasVideo": true,
    "id": "3aefb91a-be3e-42c1-b7cb-47bc80bebce5",
    "isLocal": false,
    "legacyId": "63389bb8-0aee-4da3-8fca-4f55657c3c53",
    "localizedTitle": undefined,
    "provider": undefined,
    "routeHash": "765d2c7cb80ce7405011db603ae38db6",
    "title": "Valdrôme 2 - Col du Fays, Col de Rossas ",
    "videoFormat": "mp4",
  },
  {
    "category": "Free",
    "country": "FR",
    "distance": 27205.058559774185,
    "elevation": 135.99999999999972,
    "hasGpx": false,
    "hasVideo": true,
    "id": "b69e2524-e5d1-4945-97ac-7c6b80836ff3",
    "isLocal": false,
    "legacyId": "8dbfa4c8-be36-4068-9ca1-cc9dcbfd42d7",
    "localizedTitle": undefined,
    "provider": undefined,
    "routeHash": "c62b855ca6769a53c4cc47ecd01ac276",
    "title": "Valdrôme 3 - Les Claps de Luc",
    "videoFormat": "mp4",
  },
  {
    "category": undefined,
    "country": "ES",
    "distance": 10757,
    "elevation": 628.2219999999998,
    "hasGpx": false,
    "hasVideo": true,
    "id": "fe92483905ba4a26bce4db89714ddded",
    "isLocal": true,
    "legacyId": "ad33afde-f865-419a-91ea-d09d426f07cb",
    "localizedTitle": {
      "de": "Sa Calobra",
      "en": "Sa Calobra",
    },
    "provider": undefined,
    "routeHash": "b1113dcb6b39bc44a0f0445d6b642eb7",
    "title": "SaCalobra",
    "videoFormat": "mp4",
  },
  {
    "category": undefined,
    "country": "DE",
    "distance": 8297.890839807143,
    "elevation": 25.22,
    "hasGpx": false,
    "hasVideo": true,
    "id": "06e11e6f-4517-48a7-8fd5-f8ba08b8f282",
    "isLocal": true,
    "legacyId": "021dd6dd-c08b-431f-a888-2cd3183b1911",
    "localizedTitle": undefined,
    "provider": undefined,
    "routeHash": "45ea2f828ad18b38b9c4e43c2dd18964",
    "title": "Holzweiler-Keyenberg",
    "videoFormat": "mp4",
  },
  {
    "category": undefined,
    "country": "de",
    "distance": 94259,
    "elevation": 811.0206727770162,
    "hasGpx": false,
    "hasVideo": true,
    "id": "17c4fe5b9a824e9aa450832f57a34fed",
    "isLocal": true,
    "legacyId": "ad19439c-5768-4e70-8cd1-d1b2e34caa9e",
    "localizedTitle": undefined,
    "provider": undefined,
    "routeHash": "d0d79efb023e9ff89aa8fa5b05845ed5",
    "title": "ChallengeRoth",
    "videoFormat": "mp4",
  },
  {
    "category": undefined,
    "country": "FR",
    "distance": 25647.042009507146,
    "elevation": 539.8,
    "hasGpx": false,
    "hasVideo": true,
    "id": "e1bd3245-cb2b-483a-9f28-237713445540",
    "isLocal": true,
    "legacyId": "7f78dfe9-7664-41a4-8028-74a5038dbcc7",
    "localizedTitle": undefined,
    "provider": undefined,
    "routeHash": "c3e5d6059cb517b26bfe161b36d4d72f",
    "title": "Col de Grimone - 1 - Gorges des Gâts",
    "videoFormat": "mp4",
  },
  {
    "category": undefined,
    "country": undefined,
    "distance": 24620,
    "elevation": 511.3758658286927,
    "hasGpx": false,
    "hasVideo": true,
    "id": "20210035b247e3c89d644cda39b0f271",
    "isLocal": true,
    "legacyId": "f2481125-d914-4069-ab3a-14bb97ab7fd9",
    "localizedTitle": undefined,
    "provider": undefined,
    "routeHash": "a52774c15904541d289361bce96719d9",
    "title": "Ring",
    "videoFormat": "mp4",
  },
  {
    "category": undefined,
    "country": "de",
    "distance": 24326,
    "elevation": 473.3271493465682,
    "hasGpx": false,
    "hasVideo": true,
    "id": "0d58609a1ad14d6ba12d69ffe9816633",
    "isLocal": true,
    "legacyId": "c3958847-7ec8-407c-91c0-31c27b5e72c2",
    "localizedTitle": undefined,
    "provider": undefined,
    "routeHash": "619e67811e63fd32e868c3f72330f3af",
    "title": "Rad-am-Ring",
    "videoFormat": "mp4",
  },
  {
    "category": undefined,
    "country": undefined,
    "distance": 40100,
    "elevation": 323.02932357788086,
    "hasGpx": false,
    "hasVideo": true,
    "id": "cc566cab8fb942919a7e011d540b0164",
    "isLocal": true,
    "legacyId": "6722a060-b1fc-4d72-b9e9-97a7554b7388",
    "localizedTitle": undefined,
    "provider": undefined,
    "routeHash": "8b79920fa7a4cfa6b15c6c5a3c92b61f",
    "title": "Aigen",
    "videoFormat": "mp4",
  },
  {
    "category": undefined,
    "country": undefined,
    "distance": 3801.452188724582,
    "elevation": 35.19111323845895,
    "hasGpx": true,
    "hasVideo": false,
    "id": "5c7a6ae31b1bef04c5854cb3",
    "isLocal": false,
    "legacyId": undefined,
    "localizedTitle": undefined,
    "provider": undefined,
    "routeHash": undefined,
    "title": "Sydney Opera House and Botanic Garden",
  },
  {
    "category": undefined,
    "country": undefined,
    "distance": 58144.65148583883,
    "elevation": 924.4000129702023,
    "hasGpx": true,
    "hasVideo": false,
    "id": "5c9625c44fb4dd03f6d67ec8",
    "isLocal": false,
    "legacyId": undefined,
    "localizedTitle": undefined,
    "provider": undefined,
    "routeHash": undefined,
    "title": "Sarti-Toroni-Sarti",
  },
  {
    "category": undefined,
    "country": undefined,
    "distance": 20019.063093828154,
    "elevation": 54.70924429516317,
    "hasGpx": true,
    "hasVideo": false,
    "id": "5cd7f5d8906ac87be4cb9fa8",
    "isLocal": false,
    "legacyId": undefined,
    "localizedTitle": undefined,
    "provider": undefined,
    "routeHash": undefined,
    "title": "Seaside Vancouver",
  },
  {
    "category": undefined,
    "country": undefined,
    "distance": 10129.26255082863,
    "elevation": 139.41287283607005,
    "hasGpx": true,
    "hasVideo": false,
    "id": "5d794472049b4e39617b4cfb",
    "isLocal": false,
    "legacyId": undefined,
    "localizedTitle": undefined,
    "provider": undefined,
    "routeHash": undefined,
    "title": "Malaga City Tour",
  },
  {
    "category": undefined,
    "country": undefined,
    "distance": 97061,
    "elevation": 0,
    "hasGpx": true,
    "hasVideo": false,
    "id": "600ff63b49d3f19eaa6398aa",
    "isLocal": false,
    "legacyId": undefined,
    "localizedTitle": undefined,
    "provider": undefined,
    "routeHash": undefined,
    "title": "Ventoux - Malaucene",
  },
  {
    "category": undefined,
    "country": undefined,
    "distance": 40621,
    "elevation": 0,
    "hasGpx": true,
    "hasVideo": false,
    "id": "6010012941991a02471754cb",
    "isLocal": false,
    "legacyId": undefined,
    "localizedTitle": undefined,
    "provider": undefined,
    "routeHash": undefined,
    "title": "Würzjoch ",
  },
  {
    "category": undefined,
    "country": undefined,
    "distance": 87035.68290882153,
    "elevation": 229.8657380688822,
    "hasGpx": true,
    "hasVideo": false,
    "id": "601003e141991abc5c175e0d",
    "isLocal": false,
    "legacyId": undefined,
    "localizedTitle": undefined,
    "provider": undefined,
    "routeHash": undefined,
    "title": "Captain  Cook Highway",
  },
  {
    "category": "personal",
    "country": undefined,
    "distance": 116009.98233245034,
    "elevation": 2083.7787620000013,
    "hasGpx": true,
    "hasVideo": false,
    "id": "20b1ba7a-93c6-4ce3-aafa-f9d24325c7be",
    "isLocal": true,
    "legacyId": undefined,
    "localizedTitle": undefined,
    "provider": undefined,
    "routeHash": "3cfd4fb14cf9eee2c1b9eb5b0106393e",
    "title": "Visiting Jeroen - Part 14",
  },
  {
    "category": "personal",
    "country": undefined,
    "distance": 120734.21308904006,
    "elevation": 1409.4124139999992,
    "hasGpx": true,
    "hasVideo": false,
    "id": "9b779bbc-7a20-44f9-b490-76eecac34ee9",
    "isLocal": true,
    "legacyId": undefined,
    "localizedTitle": undefined,
    "provider": undefined,
    "routeHash": "fc467f1f6ccf90f18a6baa839175f640",
    "title": "Visiting Jeroen - Part 15",
  },
]
`;
